File: 25lq/solution/java/I_链.java
import java.util.*;
import java.io.*;
class Reader {
    private BufferedReader br;
    private StringTokenizer st;

    public Reader() {
        br = new BufferedReader(new InputStreamReader(System.in));
    }

    String next() {
        while (st == null || !st.hasMoreElements()) {
            try {
                st = new StringTokenizer(br.readLine());
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
        return st.nextToken();
    }

    int nextInt() {
        return Integer.parseInt(next());
    }

    long nextLong() {
        return Long.parseLong(next());
    }

    void close() throws IOException {
        br.close();
    }
}

class Writer {
    private BufferedWriter bw;

    public Writer() {
        bw = new BufferedWriter(new OutputStreamWriter(System.out));
    }

    void print(String s) throws IOException {
        bw.write(s);
    }

    void println(String s) throws IOException {
        bw.write(s);
        bw.newLine();
    }

    void flush() throws IOException {
        bw.flush();
    }

    void close() throws IOException {
        bw.close();
    }
}

public class Main {
    public static boolean isChain(int n, int m, List<int[]> edges) {
        if (n == 1) {
            return m == 0;
        }
        if (m != n - 1) { // 链必须有n-1条边
            return false;
        }

        // 构建邻接列表
        List<List<Integer>> matrix = new ArrayList<>();
        for (int i = 0; i <= n; i++) {
            matrix.add(new ArrayList<>());
        }
        for (int[] edge : edges) {
            matrix.get(edge[0]).add(edge[1]);
            matrix.get(edge[1]).add(edge[0]);
        }

        // 检查度数并统计度数为1的点
        int degreeOne = 0;
        for (int i = 1; i <= n; i++) {
            int deg = matrix.get(i).size();
            if (deg > 2) {
                return false;
            }
            if (deg == 1) {
                degreeOne++;
            }
        }

        // 链必须有2个端点
        if (degreeOne != 2) {
            return false;
        }

        // 找到起点（度数为1的点）
        int start = 0;
        for (int i = 1; i <= n; i++) {
            if (matrix.get(i).size() == 1) {
                start = i;
                break;
            }
        }

        // DFS检查连通性和路径
        boolean[] visited = new boolean[n + 1];
        int pathLen = 0;
        int currNode = start;
        int prevNode = 0;

        while (currNode != 0) {
            visited[currNode] = true;
            pathLen++;

            int nextNode = 0;
            for (int neighbor : matrix.get(currNode)) {
                if (neighbor != prevNode && !visited[neighbor]) {
                    nextNode = neighbor;
                    break;
                }
            }

            prevNode = currNode;
            currNode = nextNode;
        }

        return pathLen == n;
    }

    public static void main(String[] args) throws IOException {
        Reader reader = new Reader();
        Writer writer = new Writer();
        int ntc = reader.nextInt();
        for (int i = 0; i < ntc; i++) {
            int n = reader.nextInt();
            int m = reader.nextInt();
            List<int[]> edges = new ArrayList<>();
            for (int j = 0; j < m; j++) {
                int u = reader.nextInt();
                int v = reader.nextInt();
                edges.add(new int[]{u, v});
            }
            writer.println(isChain(n, m, edges) ? "Yes" : "No");
        }
        writer.flush();
        reader.close();
        writer.close();
    }
}

File: 25lq/solution/I_链.py
from collections.abc import Mapping


def is_chain(n:int, m:int, edges: Mapping) -> bool:
    if n == 1:
        return m == 0
    if m != n-1:  # 链必须有n-1条边
        return False
    
    # 构建邻接矩阵
    matrix = [[] for _ in range(n+1)]
    for u, v in edges:
        matrix[u].append(v)
        matrix[v].append(u)
    
    # 检查度数并统计度数为1的点
    degree_one = 0
    for i in range(1, n+1):
        deg = len(matrix[i])
        if deg > 2:
            return False
        if deg == 1:
            degree_one += 1
    
    # 链必须有2个端点
    if degree_one != 2:
        return False
    
    # 找到起点（度数为1的点）
    for i in range(1, n+1):
        if len(matrix[i]) == 1:
            start = i
            break
    
    # DFS检查连通性和路径
    visited = [False] * (n+1)
    path_len = 0
    curr_node = start
    prev_node = 0
    
    while curr_node != 0:
        visited[curr_node] = True
        path_len += 1
        
        next_node = 0
        for neighbor in matrix[curr_node]:
            if neighbor != prev_node and not visited[neighbor]:
                next_node = neighbor
                break
        
        prev_node = curr_node
        curr_node = next_node
    
    return path_len == n

ntc = int(input())
for _ in range(ntc):
    n, m = map(int, input().split())
    edges = []
    for _ in range(m):
        u, v = map(int, input().split())
        edges.append((u, v))
    print("Yes" if is_chain(n, m, edges) else "No")
